In [1]:
import datetime

import numpy as np
import pandas as pd
import pandas_datareader as pdr  # IF NECESSARY, from terminal: pip install pandas_datareader
from numpy.random import default_rng

pd.set_option("display.max_rows", 10)  # display option for pandas
# more here: https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html

## A quick hit of Numpy

In [2]:
# create a random vector 
# every run of this --> diff #s
# see 3.2.2.2 in the textbook for why, 
# and how to prevent

rg = default_rng()
myray = rg.standard_normal(5)
print("myray:", myray) 

myray: [-0.29433307  0.75239082  0.05300588  0.3087425   0.56638009]


In [3]:
# q1 - indexing, pick the odd elements
myray[1::2]

array([0.75239082, 0.3087425 ])

Booleans arrays: Asking a logic test on an array, returns the array, where each element has been tested against that logic and converted to the boolean answer.

In [4]:
# q2 - (a) boolean array + (b) "masking"

# (a)
myray > 0

# (b)
mask = myray > 0
myray[mask] 
# or myray[myray>0]

array([0.75239082, 0.05300588, 0.3087425 , 0.56638009])

## The main event - Pandas

Vocab
- series
- index 
- dataframe
- columns and names
- rows and index 
- multiindex 

In [5]:
start = datetime.datetime(2017, 1, 1) # you can specify start and end dates this way
end = datetime.datetime(2021, 1, 27)
macro_df = pdr.data.DataReader(['GDP','CPIAUCSL','UNRATE'], 'fred', start, end)


In [6]:
# preview the dataframe
macro_df

GDP  CPIAUCSL  UNRATE
DATE                                   
2017-01-01  19153.912   243.618     4.7
2017-02-01        NaN   244.006     4.6
2017-03-01        NaN   243.892     4.4
2017-04-01  19322.920   244.193     4.4
2017-05-01        NaN   244.004     4.4
...               ...       ...     ...
2020-09-01        NaN   260.190     7.9
2020-10-01  21477.597   260.352     6.9
2020-11-01        NaN   260.721     6.7
2020-12-01        NaN   261.564     6.7
2021-01-01  22038.226   262.200     6.4

[49 rows x 3 columns]

In [7]:
# shape
macro_df.shape # no paren, bc shape is attribute, not method

(49, 3)

In [8]:
# variable types
macro_df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 49 entries, 2017-01-01 to 2021-01-01
Freq: MS
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   GDP       17 non-null     float64
 1   CPIAUCSL  49 non-null     float64
 2   UNRATE    49 non-null     float64
dtypes: float64(3)
memory usage: 1.5 KB


In [9]:
# look at top X rows
macro_df.head(3)

GDP  CPIAUCSL  UNRATE
DATE                                   
2017-01-01  19153.912   243.618     4.7
2017-02-01        NaN   244.006     4.6
2017-03-01        NaN   243.892     4.4

In [10]:
# look at bottom X rows
macro_df.tail(3)

GDP  CPIAUCSL  UNRATE
DATE                                   
2020-11-01        NaN   260.721     6.7
2020-12-01        NaN   261.564     6.7
2021-01-01  22038.226   262.200     6.4

In [11]:
# grab one variable
macro_df['GDP'] # bracket, string var name

DATE
2017-01-01    19153.912
2017-02-01          NaN
2017-03-01          NaN
2017-04-01    19322.920
2017-05-01          NaN
                ...    
2020-09-01          NaN
2020-10-01    21477.597
2020-11-01          NaN
2020-12-01          NaN
2021-01-01    22038.226
Freq: MS, Name: GDP, Length: 49, dtype: float64

In [12]:
# grab two (or more) variables
macro_df[['GDP','UNRATE']] # TWO bracket (list of var names)

GDP  UNRATE
DATE                         
2017-01-01  19153.912     4.7
2017-02-01        NaN     4.6
2017-03-01        NaN     4.4
2017-04-01  19322.920     4.4
2017-05-01        NaN     4.4
...               ...     ...
2020-09-01        NaN     7.9
2020-10-01  21477.597     6.9
2020-11-01        NaN     6.7
2020-12-01        NaN     6.7
2021-01-01  22038.226     6.4

[49 rows x 2 columns]

In [13]:
# see column names
macro_df.columns

Index(['GDP', 'CPIAUCSL', 'UNRATE'], dtype='object')

In [14]:
# change column names
macro_df.columns = ['G','C','U']
macro_df.columns

Index(['G', 'C', 'U'], dtype='object')

In [15]:
# lets change the column names back
macro_df.columns = ['GDP', 'CPIAUCSL', 'UNRATE']

In [16]:
# see index
macro_df.index

DatetimeIndex(['2017-01-01', '2017-02-01', '2017-03-01', '2017-04-01',
               '2017-05-01', '2017-06-01', '2017-07-01', '2017-08-01',
               '2017-09-01', '2017-10-01', '2017-11-01', '2017-12-01',
               '2018-01-01', '2018-02-01', '2018-03-01', '2018-04-01',
               '2018-05-01', '2018-06-01', '2018-07-01', '2018-08-01',
               '2018-09-01', '2018-10-01', '2018-11-01', '2018-12-01',
               '2019-01-01', '2019-02-01', '2019-03-01', '2019-04-01',
               '2019-05-01', '2019-06-01', '2019-07-01', '2019-08-01',
               '2019-09-01', '2019-10-01', '2019-11-01', '2019-12-01',
               '2020-01-01', '2020-02-01', '2020-03-01', '2020-04-01',
               '2020-05-01', '2020-06-01', '2020-07-01', '2020-08-01',
               '2020-09-01', '2020-10-01', '2020-11-01', '2020-12-01',
               '2021-01-01'],
              dtype='datetime64[ns]', name='DATE', freq='MS')

In [17]:
# reset_index() and set_index()
macro_df.reset_index()  # turns index into var

DATE        GDP  CPIAUCSL  UNRATE
0  2017-01-01  19153.912   243.618     4.7
1  2017-02-01        NaN   244.006     4.6
2  2017-03-01        NaN   243.892     4.4
3  2017-04-01  19322.920   244.193     4.4
4  2017-05-01        NaN   244.004     4.4
..        ...        ...       ...     ...
44 2020-09-01        NaN   260.190     7.9
45 2020-10-01  21477.597   260.352     6.9
46 2020-11-01        NaN   260.721     6.7
47 2020-12-01        NaN   261.564     6.7
48 2021-01-01  22038.226   262.200     6.4

[49 rows x 4 columns]

In [18]:
macro_df.reset_index().set_index('DATE') # turn var into index

GDP  CPIAUCSL  UNRATE
DATE                                   
2017-01-01  19153.912   243.618     4.7
2017-02-01        NaN   244.006     4.6
2017-03-01        NaN   243.892     4.4
2017-04-01  19322.920   244.193     4.4
2017-05-01        NaN   244.004     4.4
...               ...       ...     ...
2020-09-01        NaN   260.190     7.9
2020-10-01  21477.597   260.352     6.9
2020-11-01        NaN   260.721     6.7
2020-12-01        NaN   261.564     6.7
2021-01-01  22038.226   262.200     6.4

[49 rows x 3 columns]

In [19]:
# grab some rows (by position)
macro_df[4:8]

GDP  CPIAUCSL  UNRATE
DATE                                   
2017-05-01        NaN   244.004     4.4
2017-06-01        NaN   244.163     4.3
2017-07-01  19558.693   244.243     4.3
2017-08-01        NaN   245.183     4.4

In [20]:
# grab some rows (by value)
mask = macro_df['UNRATE'] > 6
macro_df[mask]

GDP  CPIAUCSL  UNRATE
DATE                                   
2020-04-01  19477.444   256.094    14.7
2020-05-01        NaN   255.944    13.2
2020-06-01        NaN   257.217    11.0
2020-07-01  21138.574   258.543    10.2
2020-08-01        NaN   259.580     8.4
2020-09-01        NaN   260.190     7.9
2020-10-01  21477.597   260.352     6.9
2020-11-01        NaN   260.721     6.7
2020-12-01        NaN   261.564     6.7
2021-01-01  22038.226   262.200     6.4

In [21]:
# create a variable
# (2 ways)
macro_df['HIGH'] = macro_df['UNRATE'] > 6
macro_df.assign(low = macro_df['UNRATE'] < 4)
macro_df 

# notice that "low" isn't in the df!
# see 3.2.4 for why
# to permanently save it, add 'df = ' to the line:
macro_df = macro_df.assign(low = macro_df['UNRATE'] < 4)


## EDA

Stop here. Back to the lecture. 

### Part 1

Q0: Do each of the [EDA golden rules for initial data exploration](https://ledatascifi.github.io/ledatascifi-2022/content/03/02e_eda_golden.html) and write down your observations.
- **Important: What is the "key" or "unit level" that this database describes?** 
    - [This is discussed in 3.2.2.2 with examples](https://ledatascifi.github.io/ledatascifi-2022/content/03/02b_pandasVocab.html#the-shape-of-data)
    - The "key" levels in databases we will look at are often increments of time in the data, the type of entity describe (e.g. firm, person, state, country, industry), and combinations of entity and time. 
    - E.g. "firm" level, "firm-year" level

Q1: What is the second series above?

Q2: What is the frequency of the series?

Q3: What is the average ANNUAL GDP, based on the data?

### Some answers
- Q0: What have we learned about the data? Anything to keep track of? (GDP is annual, others are quarterly)
- Q1: Inflation (CPI)
- Q2: Quarterly, but GDP is only annual
- Q3: 20,630 (trillion)

In [22]:
# do your work below this function

def insufficient_but_starting_eda(df,cat_vars_list=None):
    '''
    
    Parameters
    ----------
    df : DATAFRAME
    cat_vars_list : LIST, optional
        A list of strings containing variable names in the dataframe
        for variables where you want to see the number of unique values
        and the 10 most common values. Likely used for categorical values.

    Returns
    -------
    None. It simply prints.
    
    Description
    -------    
    This function will print a MINIMUM amount of info about a new dataframe. 
    
    You should ****look**** at all this output below and consider the data
    exploration and cleaning questions from 
    https://ledatascifi.github.io/ledatascifi-2021/content/03/02e_eda_golden.html#member
    
    Also LOOK at more of the data manually. 
    
    Then write up anything notable you observe.
    
    TIP: put this function in your codebook to reuse easily.
    
    PROTIP: Improve this function (better outputs, better formatting).
    
    FEATURE REQUEST: optionally print the nunique and top 10 values under the describe matrix
    
    FEATURE REQUEST: optionally print more stats (percentiles)
    
    '''
    print(df.head(),  '\n---')
    print(df.tail(),  '\n---')
    print(df.columns, '\n---')
    print("The shape is: ",df.shape, '\n---')
    print("Info:",df.info(), '\n---') # memory usage, name, dtype, and # of non-null obs (--> # of missing obs) per variable
    print(df.describe(), '\n---') # summary stats, and you can customize the list!
    if cat_vars_list != None:
        for var in cat_vars_list:
            print(var,"has",df[var].nunique(),"values and its top 10 most common are:")
            print(df[var].value_counts().head(10), '\n---')
        
insufficient_but_starting_eda(macro_df,['UNRATE'])  

                  GDP  CPIAUCSL  UNRATE   HIGH    low
DATE                                                 
2017-01-01  19153.912   243.618     4.7  False  False
2017-02-01        NaN   244.006     4.6  False  False
2017-03-01        NaN   243.892     4.4  False  False
2017-04-01  19322.920   244.193     4.4  False  False
2017-05-01        NaN   244.004     4.4  False  False 
---
                  GDP  CPIAUCSL  UNRATE  HIGH    low
DATE                                                
2020-09-01        NaN   260.190     7.9  True  False
2020-10-01  21477.597   260.352     6.9  True  False
2020-11-01        NaN   260.721     6.7  True  False
2020-12-01        NaN   261.564     6.7  True  False
2021-01-01  22038.226   262.200     6.4  True  False 
---
Index(['GDP', 'CPIAUCSL', 'UNRATE', 'HIGH', 'low'], dtype='object') 
---
The shape is:  (49, 5) 
---
<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 49 entries, 2017-01-01 to 2021-01-01
Freq: MS
Data columns (total 5 columns):
 #   Colu

In [23]:
insufficient_but_starting_eda(macro_df,['UNRATE'])  

                  GDP  CPIAUCSL  UNRATE   HIGH    low
DATE                                                 
2017-01-01  19153.912   243.618     4.7  False  False
2017-02-01        NaN   244.006     4.6  False  False
2017-03-01        NaN   243.892     4.4  False  False
2017-04-01  19322.920   244.193     4.4  False  False
2017-05-01        NaN   244.004     4.4  False  False 
---
                  GDP  CPIAUCSL  UNRATE  HIGH    low
DATE                                                
2020-09-01        NaN   260.190     7.9  True  False
2020-10-01  21477.597   260.352     6.9  True  False
2020-11-01        NaN   260.721     6.7  True  False
2020-12-01        NaN   261.564     6.7  True  False
2021-01-01  22038.226   262.200     6.4  True  False 
---
Index(['GDP', 'CPIAUCSL', 'UNRATE', 'HIGH', 'low'], dtype='object') 
---
The shape is:  (49, 5) 
---
<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 49 entries, 2017-01-01 to 2021-01-01
Freq: MS
Data columns (total 5 columns):
 #   Colu

## Part 2

Q4: Download the annual *real* gdp from 1960 to 2018 from FRED and compute the average annual percent change

In [24]:
# DL raw data
start = datetime.datetime(1960, 1, 1)
end = datetime.datetime(2018, 12, 31)

In [25]:
# chaining methods
(
    # get data
    pdr.data.DataReader('GDPCA', 'fred', start, end)
    
    # create variable: yoy change
    .assign(yoy_growth = lambda x: x['GDPCA'].pct_change())
        
    # avg annual gdp grow last 50 years
    ['yoy_growth'].mean()
)

0.03067733411159815

Q5: Compute the average gdp percent change within *each decade*


In [26]:
start = datetime.datetime(1960, 1, 1)
end = datetime.datetime(2018, 12, 31)
realGDP_df = pdr.data.DataReader('GDPCA', 'fred', start, end)
realGDP_df

GDPCA
DATE                 
1960-01-01   3262.061
1961-01-01   3345.690
1962-01-01   3550.684
1963-01-01   3705.318
1964-01-01   3918.791
...               ...
2014-01-01  16932.051
2015-01-01  17390.295
2016-01-01  17680.274
2017-01-01  18079.084
2018-01-01  18606.787

[59 rows x 1 columns]

In [27]:
# (realGDP_df.index.year//10)*10
realGDP_df['dec']=(realGDP_df.index.year//10)*10
realGDP_df['perc'] = realGDP_df['GDPCA'].pct_change()

#          for each dec, get the perc growth, and avg
realGDP_df.groupby('dec')    ['perc']       .mean()

dec
1960    0.047426
1970    0.032352
1980    0.031240
1990    0.032261
2000    0.019175
2010    0.022464
Name: perc, dtype: float64

In [28]:
(
    pdr.data.DataReader('GDPCA', 'fred', 1960, 2019)
    
    .assign(dec = lambda x: (x.index.year//10)*10,
           real_gdp_pct = lambda x: x['GDPCA'].pct_change())
    
    .groupby('dec')    ['real_gdp_pct']       .mean()
)    

dec
1960    0.047426
1970    0.032352
1980    0.031240
1990    0.032261
2000    0.019175
2010    0.022506
Name: real_gdp_pct, dtype: float64

## Part 3

First, I'll load January data on unemployment, the Case-Shiller housing index, and median household income in three states (CA/MI/PA). 

In [29]:
# do your work here

# let's pseudo code

# q6 
# get decade variable
# get annual change in unemploy for each state
# average unemploy for each state within decade

# q7
# get decade variable
# get annual pct change in house price and income for each state
# average those for each state within decade

# HEY! those are similar - let's combine q6 and q7:

# get decade variable
# get annual change in unemploy for each state
# get annual pct change in house price and income for each state
# average unemploy for each state within decade

In [30]:
# LOAD DATA AND CONVERT TO ANNUAL

start = 1990 # pandas datareader can infer these are years
end = 2018
macro_data = pdr.data.DataReader(['CAUR','MIUR','PAUR', # unemployment 
                                  'LXXRSA','DEXRSA','WDXRSA', # case shiller index in LA, Detroit, DC (no PA  available!)
                                  'MEHOINUSCAA672N','MEHOINUSMIA672N','MEHOINUSPAA672N'], #  
                                 'fred', start, end)
macro_data = macro_data.resample('Y').first() # get's the first observation for each variable in a given year

# CLEAN UP THE FORMATING SOMEWHAT

macro_data.index = macro_data.index.year
print("\n\n DATA BEFORE FORMATTING: \n\n")
print(macro_data[:20]) # see how the data looks now? ugly variable names, but its an annual dataset at least
macro_data.columns=pd.MultiIndex.from_tuples([
    ('Unemployment','CA'),('Unemployment','MI'),('Unemployment','PA'),
    ('HouseIdx','CA'),('HouseIdx','MI'),('HouseIdx','PA'),
    ('MedIncome','CA'),('MedIncome','MI'),('MedIncome','PA')
    ])
print("\n\n DATA AFTER FORMATTING: \n\n")
print(macro_data[:20]) # this is a dataset that is "wide", and now 
                       # the column variable names have 2 levels - var name, 
                       # and unit/state that variable applies to




 DATA BEFORE FORMATTING: 


      CAUR  MIUR  PAUR      LXXRSA      DEXRSA      WDXRSA  MEHOINUSCAA672N  \
DATE                                                                          
1990   5.2   7.7   5.2  100.471193         NaN   93.362855          64124.0   
1991   7.1   8.8   6.5   95.569015   58.420806   89.706871          62568.0   
1992   8.6   9.5   7.4   92.786926   59.748947   88.573807          63297.0   
1993   9.8   7.6   7.2   85.246295   61.564205   89.065118          60272.0   
1994   9.3   7.2   6.7   77.395052   64.526663   88.988467          61219.0   
...    ...   ...   ...         ...         ...         ...              ...   
2005   5.8   7.2   5.2  221.471356  123.264828  210.799936          68766.0   
2006   5.0   6.8   4.8  268.208820  126.872688  250.272524          71194.0   
2007   4.9   7.0   4.4  270.804867  118.163903  241.411115          69750.0   
2008   5.9   7.2   4.8  226.111882  100.378578  215.656546          68712.0   
2009   9.9  10.9   7.0

Q6: for each decade and state, report the average annual CHANGE (level, not percent) in unemployment

Q7: for each decade and state, report the average annual PERCENT CHANGE in house prices and household income

### Method 1: Brute force, keep dataset "shaped" as is

A small challenge: I think that this part of the code below can be collapsed to a single line of code.
```python
macro_data['ca_une_diff'] = macro_data[ ('Unemployment','CA') ].diff()
macro_data['mi_une_diff'] = macro_data[ ('Unemployment','MI') ].diff()
macro_data['pa_une_diff'] = macro_data[ ('Unemployment','PA') ].diff()
```
Please let me know if you pull it off!

In [31]:
# get decade variable

macro_data['dec'] = (macro_data.index//10)*10

# get annual change in unemploy for each state

macro_data['ca_une_diff'] = macro_data[ ('Unemployment','CA') ].diff()
macro_data['mi_une_diff'] = macro_data[ ('Unemployment','MI') ].diff()
macro_data['pa_une_diff'] = macro_data[ ('Unemployment','PA') ].diff()

# get annual pct change in house price and income for each state

macro_data['pa_house_pct'] = macro_data[ ('HouseIdx','PA') ].pct_change()*100
macro_data['ca_house_pct'] = macro_data[ ('HouseIdx','CA') ].pct_change()*100
macro_data['mi_house_pct'] = macro_data[ ('HouseIdx','MI') ].pct_change()*100

macro_data['pa_inc_pct'] = macro_data[ ('MedIncome','PA') ].pct_change()*100
macro_data['ca_inc_pct'] = macro_data[ ('MedIncome','CA') ].pct_change()*100
macro_data['mi_inc_pct'] = macro_data[ ('MedIncome','MI') ].pct_change()*100

# a setting to reduce decimals in pandas tables

pd.set_option('display.float_format', '{:,.2f}'.format)

# average our vars for each state within decade
(
    macro_data.groupby('dec') .mean()
    # this is optional adjustment to the output
    .T .tail(9) 
)    

C:\Users\DonsLaptop\anaconda3\lib\site-packages\pandas\core\generic.py:4150: PerformanceWarning: dropping on a non-lexsorted multi-index without a level parameter may impact performance.
  obj = obj._drop_axis(labels, axis, level=level, errors=errors)


dec            1990  2000  2010
ca_une_diff    0.04  0.43 -0.61
mi_une_diff   -0.41  0.69 -0.71
pa_une_diff   -0.08  0.25 -0.24
pa_house_pct   0.12  7.22  2.82
ca_house_pct  -0.71  7.52  5.84
mi_house_pct   5.92 -1.20  5.15
pa_inc_pct     0.64 -0.03  1.54
ca_inc_pct     0.69  0.01  0.84
mi_inc_pct     2.54 -2.42  1.42

In [32]:
macro_data.iloc[:10]

Unemployment           HouseIdx             MedIncome            \
               CA   MI   PA       CA    MI    PA        CA        MI   
DATE                                                                   
1990         5.20 7.70 5.20   100.47   NaN 93.36 64,124.00 57,665.00   
1991         7.10 8.80 6.50    95.57 58.42 89.71 62,568.00 59,693.00   
1992         8.60 9.50 7.40    92.79 59.75 88.57 63,297.00 58,517.00   
1993         9.80 7.60 7.20    85.25 61.56 89.07 60,272.00 57,776.00   
1994         9.30 7.20 6.70    77.40 64.53 88.99 61,219.00 61,138.00   
1995         7.80 5.20 5.70    76.38 68.97 89.67 62,618.00 61,632.00   
1996         7.80 4.90 5.90    73.92 73.58 88.66 63,965.00 64,646.00   
1997         6.90 4.70 5.30    74.54 79.35 88.78 64,035.00 62,499.00   
1998         6.00 4.00 4.80    81.04 85.18 90.11 65,153.00 66,564.00   
1999         5.60 4.00 4.50    92.14 92.43 94.19 67,994.00 71,828.00   

                 dec ca_une_diff mi_une_diff pa_une_diff pa_house_pct  \
            PA                                                          
DATE                                                                    
1990 55,870.00  1990         NaN         NaN         NaN          NaN   
1991 56,440.00  1990        1.90        1.10        1.30        -3.92   
1992 54,191.00  1990        1.50        0.70        0.90        -1.26   
1993 54,827.00  1990        1.20       -1.90       -0.20         0.55   
1994 55,562.00  1990       -0.50       -0.40       -0.50        -0.09   
1995 58,413.00  1990       -1.50       -2.00       -1.00         0.77   
1996 57,516.00  1990        0.00       -0.30        0.20        -1.13   
1997 60,523.00  1990       -0.90       -0.20       -0.60         0.14   
1998 62,098.00  1990       -0.90       -0.70       -0.50         1.50   
1999 58,844.00  1990       -0.40        0.00       -0.30         4.53   

     ca_house_pct mi_house_pct pa_inc_pct ca_inc_pct mi_inc_pct  
                                                                 
DATE                                                             
1990          NaN          NaN        NaN        NaN        NaN  
1991        -4.88          NaN       1.02      -2.43       3.52  
1992        -2.91         2.27      -3.98       1.17      -1.97  
1993        -8.13         3.04       1.17      -4.78      -1.27  
1994        -9.21         4.81       1.34       1.57       5.82  
1995        -1.32         6.89       5.13       2.29       0.81  
1996        -3.22         6.68      -1.54       2.15       4.89  
1997         0.83         7.84       5.23       0.11      -3.32  
1998         8.72         7.35       2.60       1.75       6.50  
1999        13.70         8.51      -5.24       4.36       7.91

### Method 2: "Tidy" the data and reshape it

The basic idea is that instead of doing the variable creation 3 times for each variable, that if the unemployment variables were stacked on top of each other, we'd only have to do the variable creation once per variable. To do that, we need to convert the data from "wide" to tall.  

For more, see discussion in the textbook about data shapes.

**First, we should reload the data from scratch:**

In [33]:
# LOAD DATA AND CONVERT TO ANNUAL

start = 1990 # pandas datareader can infer these are years
end = 2018
macro_data = pdr.data.DataReader(['CAUR','MIUR','PAUR', # unemployment 
                                  'LXXRSA','DEXRSA','WDXRSA', # case shiller index in LA, Detroit, DC (no PA  available!)
                                  'MEHOINUSCAA672N','MEHOINUSMIA672N','MEHOINUSPAA672N'], #  
                                 'fred', start, end)
macro_data = macro_data.resample('Y').first() # get's the first observation for each variable in a given year

# CLEAN UP THE FORMATING SOMEWHAT

macro_data.index = macro_data.index.year
macro_data.columns=pd.MultiIndex.from_tuples([
    ('Unemployment','CA'),('Unemployment','MI'),('Unemployment','PA'),
    ('HouseIdx','CA'),('HouseIdx','MI'),('HouseIdx','PA'),
    ('MedIncome','CA'),('MedIncome','MI'),('MedIncome','PA')
    ])


Now we can try the "Method 2" approach:

In [34]:
(
    # reformat the data to tall with "stack"
    macro_data.stack()
    
    # to make later steps (like the groupby) a touch easier, 
    # convert year and state from the index into variables
    .reset_index().rename(columns={'level_1':'state'})
    
    # now order the dataset sensibly
    .sort_values(['state','DATE'])    
    
    # create vars <---- this is not even needed to explain this block!
    .assign(
            decade          = lambda x: 10*np.floor(x['DATE']/10).astype(int),
            unemploy_diff   = lambda x: x['Unemployment'].diff(),
            HouseIdx_pctch  = lambda x: x['HouseIdx'].pct_change()*100,
            MedIncome_pctch = lambda x: x['MedIncome'].pct_change()*100    
    )
    
    # opt A for output: pivot_table
#     .pivot_table(index='decade',
#                  columns='state',
#                  values=['unemploy_diff','HouseIdx_pctch','MedIncome_pctch'])
#     .multiply(100) # for more meaningful displays 
    
    # opt B for output: groupby 
    # also, optional advanced formatting 
    .groupby(['state','decade'])
    [['unemploy_diff','HouseIdx_pctch','MedIncome_pctch']].mean()
    .style.format({'HouseIdx_pctch': '{0:,.2f}%',
                   'MedIncome_pctch': '{0:,.2f}%',
                   'unemploy_diff': '{0:,.2f} p.p.'})
)    

## Hmmmmmmmmm

**There is a slight problem: Compare the output of Method 1 and Method 2... they give slightly different answers.**

What causes the difference? Well the data must somehow be different right before doing the output!

So let's look at the data right after you're done reworking it (and right before the output). That's always a good idea, and a hyper smart version of ABCD.

Clue: Method 2 gives different numbers for the 1990s only, and for only MI and PI, but 1990-CA is the same as Method 1.

So, let's look at the part of the dataframe for Michigan in the 1990s, plus the rows right above where Michigan starts.


In [35]:
(
# this is the same as above...
    
    # reformat the data to tall with "stack"
    macro_data.stack()
    
    # to make later steps (like the groupby) a touch easier, 
    # convert year and state from the index into variables
    .reset_index().rename(columns={'level_1':'state'})
    
    # now order the dataset sensibly
    .sort_values(['state','DATE'])    
    
    # create vars <---- this is not even needed to explain this block!
    .assign(
            decade          = lambda x: 10*np.floor(x['DATE']/10).astype(int),
            unemploy_diff   = lambda x: x['Unemployment'].diff(),
            HouseIdx_pctch  = lambda x: x['HouseIdx'].pct_change(),
            MedIncome_pctch = lambda x: x['MedIncome'].pct_change()    
    )
   
# ok this is new:    
# look at the part where CA ends and MI begins,
# and focus on the houseidx variables
.iloc[25:34,[0,1,2,-2] ]
)  

DATE state  HouseIdx  HouseIdx_pctch
75  2015    CA    228.19            0.05
78  2016    CA    243.38            0.07
81  2017    CA    256.08            0.05
84  2018    CA    274.91            0.07
1   1990    MI       NaN            0.00
4   1991    MI     58.42           -0.79
7   1992    MI     59.75            0.02
10  1993    MI     61.56            0.03
13  1994    MI     64.53            0.05

That can't possibly be right! MI doesn't have -78% housing price drop in 1991! It's because pct_change compares one row to the prior row with available data. So it compares 1991 MI to 2019 CA!

The lesson as always: ABCD!